File: pkg/core/vector/binary/binary.go
package binary

import "unsafe"

const (
	//Zero predefined constant for uint zero representation
	Zero = uint8(0)
	//One predefined constant for uint one's representation
	One = uint8(1)
)

//RoudDownBin returns 2 base number rounded down in O(log(N))
func RoudDownBin(x uint8) uint8 {
	x |= x >> 1
	x |= x >> 2
	x |= x >> 4
	return x - (x >> 1)
}

//RoudUpBin is derivative function to round up number to near power of 2 in O(log(N))
func RoudUpBin(x uint8) uint8 {
	return RoudDownBin(x) << 1
}

//IsBinRadix returns 1 in case if given number is radix of 2 and 0 otherwise in O(1)
func IsBinRadix(x uint8) uint8 {
	return (x & (x - 1)) ^ 0
}

//Invert implements bit inversion with O(log(N)) required by FFT
func Invert(x, n uint8) uint8 {
	var y uint8
	for i := Zero; i < n; i++ {
		y <<= 1
		y |= x & 1
		x >>= 1
	}
	return y
}

//LogBin finds log with  base 2 of 32 bit number with O(log(N)) complexity and without branching
func LogBin(x uint8) uint8 {
	y := BoolToInt(x > 0xF) << 2
	x >>= y
	shift := BoolToInt(x > 0x3) << 1
	x >>= shift
	y |= shift
	return x>>1 | y
}

//BoolToInt used for fast conversion bool to uint8
func BoolToInt(x bool) uint8 {
	return *(*uint8)(unsafe.Pointer(&x)) & 1
}

//IntToBool used for fast conversion from uint16 to bool
func IntToBool(x uint8) bool {
	x &= 1
	return *(*bool)(unsafe.Pointer(&x))
}

File: src/binary/binary.rs
extern crate num;

use num::Unsigned;

// 
pub fn log_bin(mut x : usize) -> usize {
	let mut i : usize = 32;
	while i > 0 {
		x |= x >> i;
		i >>= 1;
	}
	return x - (x >> 1);
}

//is_bin_radix returns 1 if given number divisible by 2 without reminder and returns 0 otherwise in O(1)
pub fn is_bin_radix <T> (x : T) -> T 
where T: PrimInt + Unsigned {
	return !(x & (x - 1));
}

//log_bin finds log with  base 2 of 64 bit number with O(log(N)) complexity
pub	fn log(mut x : usize) -> usize {
	if x == 0 {
		return 0;
	}
	let mask : [usize; 6] = [ 0xFFFFFFFF00000000, 0xFFFF0000, 0xFF00, 0xF0, 0xC, 0x2 ];
	let mut y : usize = (x & (x - 1)) & 1;
	let mut j : usize = 32;
	for i in 0..6 {
		if x & mask[i] != 0 {
			y += j;
			x >>= j;
		};
	  j >>= 1;
	}
	return y
}

//Invert implements bit inversion with O(log(N)) required by FFT
pub fn reverse(mut x : usize, len : usize) -> usize {
	let mut size : usize = len;
	let mut mask : usize = !0;         
	while size > 0 {
		size >>= 1;
  		mask ^= mask << size;
  		x = ((x >> size) & mask) | ((x << size) & !mask);
	}
	return x;
}

/*
//RoudDownBin returns 2 base number rounded down in O(log(N))
pub fn order(mut x : u16) -> u16 {
	x |= x >> 1;
	x |= x >> 2;
	x |= x >> 4;
	x |= x >> 8;
	return x - (x >> 1);
}

pub fn inverse_bit(mut x : u16) -> u16 {
	let mut mask : u16 = 1 << log_bin(x);
	mask |= mask - 1;
	return x ^ mask
}*/
